main/src/library/Nec.flix

```diff
@@ -0,0 +1,909 @@
+            case OneLeft(x)      => {let a = f(x); k(NecOne(a))}
+            case SomeLeft(x, rs) => loop(rs, ks -> {let a = f(x); k(cons(a, ks))})
+        };
+        loop(c, k -> k as & ef)
+
+    ///
+    /// Returns the result of applying `f` to every element in `c` along with that element's index.
+    ///
+    /// That is, the result is of the form: `f(x1, 0) :: f(x2, 1) :: ...`.
+    ///
+    pub def mapWithIndex(f: (a, Int32) -> b & ef, c: Nec[a]): Nec[b] & ef =
+        def loop(c1: Nec[a], i: Int32, k: Nec[b] -> Nec[b] & ef) = match viewLeft(c1) {
+            case OneLeft(x)      => {let a = f(x, i); k(NecOne(a))}
+            case SomeLeft(x, rs) => loop(rs, i+1, ks -> {let a = f(x, i); k(cons(a, ks))})
+        };
+        loop(c, 0, k -> k as & ef)
+
+    ///
+    /// Apply every function from `f` to every argument from `x` and return a Nec with all results.
+    /// For `f = f1, f2, ...` and `x = x1, x2, ...` the results appear in the order
+    /// `f1(x1), f1(x2), ..., f2(x1), f2(x2), ...`.
+    ///
+    pub def ap(f: Nec[a -> b & ef], c: Nec[a]) : Nec[b] & ef =
+        def loop(f1: Nec[a -> b & ef], k: Nec[b] -> Nec[b] & ef) = match viewLeft(f1) {
+            case OneLeft(f2)      => k(map(f2, c))
+            case SomeLeft(f2, rs) => loop(rs, ks -> k(map(f2, c) `append` ks))
+        };
+        loop(f, k -> k as & ef)
+
+    ///
+    /// Returns the result of applying `f` to every element in `c` and concatenating the results.
+    ///
+    pub def flatMap(f: a -> Nec[b] & ef, c: Nec[a]): Nec[b] & ef =
+        def loop(c1: Nec[a], k: Nec[b] -> Nec[b] & ef) = match viewLeft(c1) {
+            case OneLeft(x)      => {let a = f(x); k(a)}
+            case SomeLeft(x, rs) => loop(rs, ks -> {let a = f(x); k(append(a, ks))})
+        };
+        loop(c, k -> k as & ef)
+
+    ///
+    /// Returns the reverse of `c`.
+    ///
+    pub def reverse(c: Nec[a]): Nec[a] =
+        // Use an accumulator rather than CPS, as it will be built "naturally" in reverse order.
+        def loop(c1: Nec[a], acc: Nec[a]) = match viewLeft(c1) {
+            case OneLeft(x)      => cons(x, acc)
+            case SomeLeft(x, rs) => loop(rs, cons(x, acc))
+        };
+        // Do one step before the loop so we have some content.
+        match viewLeft(c) {
+            case OneLeft(x)      => singleton(x)
+            case SomeLeft(x, rs) => loop(rs, singleton(x))
+        }
+
+    ///
+    /// Returns `l` with every occurrence of `from` replaced by `to`.
+    ///
+    pub def replace(from: {from :: a}, to: {to :: a}, l: Nec[a]): Nec[a] with Eq[a] =
+        map(e -> if (e == from.from) to.to else e, l)
+
+    ///
+    /// Returns all permutations of `c` in lexicographical order by element indices in `c`.
+    ///
+    /// That is, `c` is the first permutation and `reverse(c)` is the last permutation.
+    ///
+    pub def permutations(c: Nec[a]): Nec[List[a]] = match viewLeft(c) {
```

Comment:
Same for `init` and `tail`

Reply:
And `subsequences` :)

Reply:
Firstly I think the output sequences (if not the sequence types) of `subsequences` and `permutations` should match the List output:

`subsequences` on a list produces a Nil:

~~~

$ List.subsequences(10 :: 11 :: 12 :: Nil) |> List.foreach(println)

10 :: 11 :: 12 :: Nil
10 :: 11 :: Nil
10 :: 12 :: Nil
10 :: Nil
11 :: 12 :: Nil
11 :: Nil
12 :: Nil
Nil

~~~

So the `Nec` version should produce a "nil" - by definition it can't produce a Nil Nec so it has to produce Lists, because it always produces at least two Lists it should return `Nec[List[a]]` - a Nec of List. It could potentially return `Nec[Chain[a]]` but `List` is the more common type and `List` is favoured by the `Reducible` class.

`Nec.permutations` which currently `Nec[List[a]]` is arguably the wrong type - the permutations are the same length as the input sequence so we know they will never be empty, hence we could have the answer type `Nec[Nec[a]]`. However, the currently type is analogue to the current type of `Nel.permutations` so that is a reason in its favour.

`Nec.init` and `Nec.tail` likewise follow `Nel` - they must allow "nil" - arguably again they could return a `Chain` rather than a `List`, but returning a `List` is compatible with the `Reducible` class.